pkg/model/mselection/heaviest.go

```diff
@@ -191,23 +245,21 @@ func (s *HeaviestSelector) OnNewSolidBundle(bndl *tangle.Bundle) {
 					panic(fmt.Sprintf("transaction not found: %v", hornet.Hash(approveeHash).Trytes()))
 				}
 
-				_, approveeORTSI = s.getTransactionRootSnapshotIndexes(cachedApproveeTx.Retain(), lsmi) // tx +1
+				_, approveeORTSI = dag.GetTransactionRootSnapshotIndexes(cachedApproveeTx.Retain(), lsmi) // tx +1
 				cachedApproveeTx.Release(true)
 			}
 
-			// if the confirmationIdx to LSMI delta of the approvee is equal or greater belowMaxDepth, the tip is invalid.
+			// if the approveeORTSI to LSMI delta of the approvee is equal or greater belowMaxDepth, the tip is invalid.
 			// "equal" is important because the next milestone would reference this transaction.
 			if lsmi-approveeORTSI >= belowMaxDepth {
-				return
+				return s.GetStats()
 			}
 		}
 	}
 
-	// TODO: when len(s.approvers) gets too large trigger a checkpoint to prevent drastic performance hits
-
 	// compute the referenced transactions
 	idx := uint(len(s.approvers))
-	it := &item{hash: bndl.GetTailHash(), index: idx, refs: bitset.New(idx + 1).Set(idx)}
+	it := &item{hash: bndl.GetTailHash(), refs: bitset.New(idx + 1).Set(idx)}
```

Comment:
I would add a comment here how this works (e.g. we expand the bitset, it references easy added approver, etc.), so we do not have to reverse-engineer this again in a couple of months :-)